src/Features/CSharp/Portable/ExtractInterface/CSharpExtractInterfaceService.cs

```diff
@@ -30,25 +31,24 @@ internal sealed class CSharpExtractInterfaceService : AbstractExtractInterfaceSe
         public CSharpExtractInterfaceService()
         {
         }
-
         protected override async Task<SyntaxNode> GetTypeDeclarationAsync(Document document, int position, TypeDiscoveryRule typeDiscoveryRule, CancellationToken cancellationToken)
         {
-            var tree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
-            var root = await tree.GetRootAsync(cancellationToken).ConfigureAwait(false);
-            var token = root.FindToken(position != tree.Length ? position : Math.Max(0, position - 1));
-            var typeDeclaration = token.GetAncestor<TypeDeclarationSyntax>();
+            var span = new TextSpan(position, 0);
+            var nodes = await document.GetRelevantNodesAsync<SyntaxNode>(span, cancellationToken).ConfigureAwait(false);
+
+            var typeDeclaration = nodes.Where(n => n is TypeDeclarationSyntax).FirstOrDefault() ?? nodes.FirstOrDefault();
 
-            if (typeDeclaration == null ||
-                typeDiscoveryRule == TypeDiscoveryRule.TypeDeclaration)
+            if (typeDeclaration == null)
             {
                 return typeDeclaration;
             }
 
-            var spanStart = typeDeclaration.Identifier.SpanStart;
-            var spanEnd = typeDeclaration.TypeParameterList != null ? typeDeclaration.TypeParameterList.Span.End : typeDeclaration.Identifier.Span.End;
-            var span = new TextSpan(spanStart, spanEnd - spanStart);
+            if (typeDiscoveryRule == TypeDiscoveryRule.TypeNameOnly)
+            {
+                return typeDeclaration.Span.IntersectsWith(position) ? typeDeclaration : null;
+            }
```

Comment:
i don't quite get this check.  it says "Type Name Only", but then supports being anywhere within the entire span of the type-decl?

Reply:
Thanks :)

Reply:
I was thinking about this wrong (Thank you to Sam for providing context and clarity!) The two sets of behaviors are for... 
- Code action should only be triggered from the declaration line. This case is represented by TypeNameOnly, because in the previous implementation it was only triggered on the name. New implementation triggers anywhere in the type definition. 
- Execution of the command is still allowed from anywhere inside. This case is represented by TypeDefinition. Any position inside of the TypeDefinition closure should return the parent type definition for this case. 

This IntersectsWith code was completely useless, this would have worked the same if it just returned `typeDeclaration` variable. I've updated the method to be more clear. An update to these variable names would probably also be good? Maybe "TypeDefinition" becomes "InsideTypeDefinition"?

Reply:
> New implementation triggers anywhere in the type definition.

Does this truly mean 'anywhere'?  For example, if i'm deep within a method somewhere, are we could to start adding 'extract interface' to the lightbulb list?

Reply:
I added a gif of the behavior to the PR description

Reply:
awesome.  the gif was super helpful :)